src/gordon_gcp/clients/gdns.py

```diff
@@ -97,47 +98,117 @@ class GDNSClient(http.AIOConnection):
-            rrset = GCPResourceRecordSet(**record)
-            records.append(rrset)
+    @staticmethod
+    def get_rrsets_as_objects(rrsets):
+        """Return a list of rrsets as GCPResourceRecordSets.
+
+        Args:
+            rrsets (list of dict): RRsets represented as dicts.
+        Returns:
+            list of :class:`GCPResourceRecordSet` objects.
+        """
+        return [GCPResourceRecordSet(**rrset) for rrset in rrsets]
+
+    def get_managed_zone(self, zone):
+        """Get the GDNS managed zone name for a DNS zone.
 
-    async def get_records_for_zone(self, zone):
-        """Get all resource record sets for a particular managed zone.
+        Google uses custom string names with specific `requirements
+        <https://cloud.google.com/dns/api/v1/managedZones#resource>`_
+        for storing records. The scheme implemented here chooses a
+        managed zone name which removes the trailing dot and replaces
+        other dots with dashes, and in the case of reverse records,
+        uses only the two most significant octets, prepended with
+        'reverse'.
+
+        Example:
+           get_managed_zone('example.com.') = 'example-com'
+           get_managed_zone('30.20.10.in-addr.arpa.) = 'reverse-20-10'
 
         Args:
-            zone (str): Desired managed zone to query.
+            zone (str): DNS zone.
         Returns:
-            list of :class:`GCPResourceRecordSet` instances.
+            str of managed zone name.
+
         """
-        url = f'{self._base_url}/managedZones/{zone}/rrsets'
-
-        # to limit the amount of data across the wire; also makes it
-        # easier to create GCPResourceRecordSet instances
-        fields = ('rrsets/name,rrsets/kind,rrsets/rrdatas,'
-                  'rrsets/type,rrsets/ttl,nextPageToken')
-        params = {
-            'fields': fields,
-        }
+        if zone.endswith('.in-addr.arpa.'):
+            return self.REVERSE_PREFIX + '-'.join(zone.split('.')[-5:-3])
+        return '-'.join(zone.split('.')[:-1])
+
+    async def get_records_for_zone(self, dns_zone, params=None):
+        """Get all resource record sets for a managed zone, using the DNS zone.
+
+        Args:
+            dns_zone (str): Desired DNS zone to query.
+            params (dict): Optional set of search parameters to pass to GDNS.
+        Returns:
+            list of dicts representing rrsets.
+        """
+        managed_zone = self.get_managed_zone(dns_zone)
+        url = f'{self._base_url}/managedZones/{managed_zone}/rrsets'
+
+        if not params:
+            params = {}
+        else:
+            # otherwise we can break things in callers because we're actually
+            # modifying the same object; see
+            # https://en.wikipedia.org/wiki/Evaluation_strategy#Call_by_sharing
+            # https://jeffknupp.com/blog/2012/11/13/is-python-callbyvalue-or-callbyreference-neither/
+            params = copy.deepcopy(params)
```

Comment:
Hmm, I'm having trouble understanding what the problem is that creating a new dictionary object is trying to solve.

What it _seems_ like this is trying to solve is the issue with `params` mutating. For instance (walking thru this mostly for my benefit to make sure I understand):

```pycon
>>> some_params = {'foo': 'bar'}
>>> def foo(params=None):
...   if not params:
...     params = {}
...   print(f'1: {params}')
...   params['a_new_key'] = 'some value'
...   print(f'2: {params}')
...

# no `params` arg given, `params` value is not mutated
>>> foo()
1: {}
2: {'a_new_key': 'some value'}
>>> foo()
1: {}
2: {'a_new_key': 'some value'}

# `params` arg given, `params` value is mutated
>>> foo(some_params)
1: {'foo': 'bar'}
2: {'foo': 'bar', 'a_new_key': 'some value'}
>>> foo(some_params)
# potentially expecting {'foo': 'bar'} here
1: {'foo': 'bar', 'a_new_key': 'some value'} 
2: {'foo': 'bar', 'a_new_key': 'some value'}
```

So then creating a new object via deep copy fixes that:

```pycon
>>> some_params = {'foo': 'bar'}
>>> def foo(params=None):
...   if not params:
...     params = {}
...   else:
...     params = copy.deepcopy(params)
...   print(f'1: {params}')
...   params['a_new_key'] = 'some value'
...   print(f'2: {params}')
...
>>> foo(some_params)
1: {'foo': 'bar'}
2: {'foo': 'bar', 'a_new_key': 'some value'}
>>> foo(some_params)
1: {'foo': 'bar'}
2: {'foo': 'bar', 'a_new_key': 'some value'}
```

But I don't know if that's a problem in our case, as when we call this coroutine method `get_records_for_zone` in the enricher, we don't re-use `params`; we always create a new object, `search_params`, each time we call the method. A simplified calling path:

```pycon
# simplification of this current method 
>>> def get_records_for_zone(params=None):
...   if not params:
...     params = {}
...   print(f'1: {params}')
...   if not 'a_new_key' in params:
...     params['a_new_key'] = 'some_value'
...   print(f'2: {params}')

# simplification of the calling method that uses the params arg
>>> def _get_matching_records():
...   search_params = {'name': 'foo.spotify.net', 'type': 'A'}
...   get_records_for_zone(search_params)

# _get_matching_records called per message received
>>> _get_matching_records()
1: {'name': 'foo.spotify.net', 'type': 'A'}
2: {'name': 'foo.spotify.net', 'type': 'A', 'a_new_key': 'some_value'}
>>> _get_matching_records()
1: {'name': 'foo.spotify.net', 'type': 'A'}
2: {'name': 'foo.spotify.net', 'type': 'A', 'a_new_key': 'some_value'}
```

*But* I recognize that I've been without Gordon context for the past 3 weeks, so I could very well be missing something. It's also friday :D

Reply:
Those top two code blocks, yes.  The problem isn't in the call from the enricher, it's from the tests.